filters/include/pcl/filters/impl/crop_hull.hpp

```diff
@@ -104,81 +84,94 @@ pcl::CropHull<PointT>::getHullCloudRange ()
-        // once a point has tested +ve for being inside one polygon, we can
-        // stop checking the others:
-        break;
+      cropBoxInlierMask.at(idx) = false;
+      if (!keepInside) {
+        indices.push_back(idx);
+      }
+      else if (extract_removed_indices_) {
+        removed_indices_->push_back(idx);
       }
     }
-    // If we're removing points *inside* the hull, only remove points that
-    // haven't been found inside any polygons
-    if (poly == hull_polygons_.size () && !crop_outside_)
-      output.push_back (input_->points[(*indices_)[index]]);
   }
-}
 
-//////////////////////////////////////////////////////////////////////////////////////////////////////////////////
-template<typename PointT> template<unsigned PlaneDim1, unsigned PlaneDim2> void 
-pcl::CropHull<PointT>::applyFilter2D (std::vector<int> &indices)
-{
-  // see comments in (PointCloud& output) overload
-  for (std::size_t index = 0; index < indices_->size (); index++)
+  for (const index_t idx : *indices_)
   {
-    std::size_t poly;
-    for (poly = 0; poly < hull_polygons_.size (); poly++)
-    {
-      if (isPointIn2DPolyWithVertIndices<PlaneDim1,PlaneDim2> (
-              input_->points[(*indices_)[index]], hull_polygons_[poly], *hull_cloud_
-         ))
-      {
-        if (crop_outside_)      
-          indices.push_back ((*indices_)[index]);
-        break;
-      }
+    if (!cropBoxInlierMask[idx]) {
+      continue;
+    }
+    // iterate over polygons faster than points because we expect this data
+    // to be, in general, more cache-local - the point cloud might be huge
+    const auto poly_it = std::find_if(
+        hull_polygons_.cbegin(), hull_polygons_.cend(),
+        [idx, this](const Vertices & poly) -> bool {
+          return isPointIn2DPolyWithVertIndices<PlaneDim1,PlaneDim2> (
+              input_->points[idx], poly, *hull_cloud_);
+        });
+    const bool found_in_polygons = (poly_it != hull_polygons_.cend());
+    if (keepInside == found_in_polygons) {
+      // valid index: either found inside and need to keep inside
+      //                     or not found inside and need to remove inside
+      indices.push_back (idx);
+    }
+    else if (extract_removed_indices_) {
+      removed_indices_->push_back (idx);
     }
-    if (poly == hull_polygons_.size () && !crop_outside_)
-      indices.push_back ((*indices_)[index]);
   }
 }
 
 //////////////////////////////////////////////////////////////////////////////////////////////////////////////////
-template<typename PointT> void 
-pcl::CropHull<PointT>::applyFilter3D (PointCloud &output)
+template<typename PointT> void
+pcl::CropHull<PointT>::applyFilter3D (pcl::Indices &indices)
 {
   // This algorithm could definitely be sped up using kdtree/octree
   // information, if that is available!
 
-  for (std::size_t index = 0; index < indices_->size (); index++)
+  const bool keepInside = crop_outside_ ^ negative_;
+
+  pcl::Indices cropInlierIndices;
```

Comment:
Let's add a note here why we're doing this since this is the core contribution of the PR.